### PR TITLE
Feature/handle game

### PR DESCRIPTION
**Pull request by meiravanidjar**

### CODE REVIEW

File: src/main/java/CheckersGame.java [hunk 1]
```diff
@@ -25,11 +42,41 @@ public User getActiveUser() {
     @Override
     public Map<Piece, List<Move>> listPossibleMoves() {
         Preconditions.checkState(status == Status.IN_PROGRESS);
-        return null;
+        Map<Piece, List<Move>> currMap = new HashMap();
```
Comment (ilyash-b) [inline src/main/java/CheckersGame.java:45]: Should be `new HashMap<>()`

File: src/main/java/CheckersGame.java [hunk 2]
```diff
@@ -25,11 +42,41 @@ public User getActiveUser() {
...
+        return currMap;
     }
 
     @Override
     public void makeMove(@NotNull User user, @NotNull Move move) {
-        // TODO
+
+        this.board.setPiece(move.end(), this.board.getPiece(move.start()));
+        this.board.setPiece(move.start(), null);
+        if (move.intermediates() != null) { this.board.setPiece(move.intermediates(), null); }
```
Comment (ilyash-b) [inline src/main/java/CheckersGame.java:67]: `setPiece(..., null)` should be `removePiece(...)`

File: src/main/java/Location.java [hunk 3]
```diff
@@ -6,4 +6,7 @@ public record Location(int x, int y) {
         Preconditions.checkArgument(x >= 0 && x < 8);
         Preconditions.checkArgument(y >= 0 && y < 8);
     }
+
+    public int getX() { return this.x; }
+    public int getY() { return this.y; }
```
Comment (ilyash-b) [inline src/main/java/Location.java:11]: Java record provides getters. Please don't add these.

File: src/main/java/Move.java [hunk 4]
```diff
@@ -1,4 +1,4 @@
 import java.util.List;
 
 // Was not in project plan
-public record Move(Location start, Location end, List<Location> intermediates) { }
+public record Move(Location start, Location end, Location intermediates) { }
```
Comment (ilyash-b) [inline src/main/java/Move.java:4]: `intermediates` supposed to hole multiple intermediate locations, not single

File: src/main/java/CheckersGame.java [hunk 5]
```diff
@@ -13,8 +14,24 @@ public class CheckersGame implements Game {
...
+            // For rows 0,1,2 color is white, skip row 3-4, then use color 5
+            if (row == 3) {
+                row = 5;
+                currColor = Piece.Color.BLACK;
+            }
+            // Go over cols and place soldiers
+            for (int col = ((row % 2) == 0) ? 0 : 1; col < 8; col = col + 2) {
+                // Place a soldier at current location with the right color
+                Location currLocation = new Location(row,col);
+                board.setPiece(currLocation, new CheckersPiece(board, currColor, currLocation));
```
Comment (ilyash-b) [inline src/main/java/CheckersGame.java:32]: I recommend that `CheckersPiece` constructor would place the piece on the board instead of `board.setPiece()` here.

File: src/main/java/CheckersGame.java [hunk 6]
```diff
@@ -25,11 +42,41 @@ public User getActiveUser() {
...
+        }
+
+        return currMap;
     }
 
     @Override
     public void makeMove(@NotNull User user, @NotNull Move move) {
-        // TODO
+
+        this.board.setPiece(move.end(), this.board.getPiece(move.start()));
```
Comment (ilyash-b) [inline src/main/java/CheckersGame.java:65]: Todo: remove opponent's pieces when appropriate

File: src/main/java/CheckersKingPiece.java [hunk 7]
```diff
@@ -1,15 +1,76 @@
...
+        }
+
         return super.listPossibleMoves();
     }
+
+    private boolean isNextCellEmpty(int directionX, int directionY){
+        if (this.board.getPiece(new Location(location.getX() + directionX, this.location.getY() + directionY)) == null){
+            return true;
+        }
+        else { return false; }
```
Comment (ilyash-b) [inline src/main/java/CheckersKingPiece.java:55]: Please replace all `if(COND) { return true; } else { return false; }` with `return COND`.

File: src/main/java/CheckersKingPiece.java [hunk 8]
```diff
@@ -1,15 +1,76 @@
...
-    public CheckersKingPiece(@NotNull Board board, @NotNull Color pieceColor) {
-        super(board, pieceColor);
+    public CheckersKingPiece(@NotNull Board board, @NotNull Color pieceColor,@NotNull Location location) {
+        super(board, pieceColor,location);
     }
 
     @Override
     public List<Move> listPossibleMoves() {
-        // TODO: actually different from regular piece
+        List<Move> moves = new ArrayList<Move>();
```
Comment (ilyash-b) [inline src/main/java/CheckersKingPiece.java:13]: Bug: `moves` is never read from. (My IDE spotted this)

File: src/main/java/CheckersKingPiece.java [hunk 9]
```diff
@@ -1,15 +1,76 @@
...
 
 public class CheckersKingPiece extends CheckersPiece {
-    public CheckersKingPiece(@NotNull Board board, @NotNull Color pieceColor) {
-        super(board, pieceColor);
+    public CheckersKingPiece(@NotNull Board board, @NotNull Color pieceColor,@NotNull Location location) {
+        super(board, pieceColor,location);
     }
 
     @Override
     public List<Move> listPossibleMoves() {
```
Comment (ilyash-b) [inline src/main/java/CheckersKingPiece.java:12]: I would expect `listPossibleMoves()` to have a recursion somewhere but it doesn't. Looks like multi-step moves are completely ignored.

File: src/main/java/Move.java (hunk 4 above)
Comment (ilyash-b) [reply src/main/java/Move.java:4]: It also looks like `intermediates` are used for the captured piece of the oponent.
* `intermediates` were intended for intermediate steps of this piece, not for captured pieces
* In both cases, it's multiple, not one

File: src/main/java/CheckersKingPiece.java [hunk 10]
```diff
@@ -1,15 +1,76 @@
...
+        }
+        else if (isAbleToCapture(-2, -2)){
+            moves.add(new Move(this.location, new Location(this.location.getX() + 2, this.location.getY() + 2),
+                      new Location(this.location.getX() - 1, this.location.getY() - 1)));
+        }
+
         return super.listPossibleMoves();
     }
+
+    private boolean isNextCellEmpty(int directionX, int directionY){
```
Comment (ilyash-b) [inline src/main/java/CheckersKingPiece.java:51]: Is there any check for being on the board or outside of the board?

File: src/main/java/CheckersPiece.java [hunk 11]
```diff
@@ -1,25 +1,90 @@
...
+        if (this.location.getY() > 0 && isRowValid(1) && isNextLeftCellEmpty()){
+            moves.add(new Move(this.location, new Location(this.location.getX() + colorMod, this.location.getY() - 1), null));
+        }
+        else if (this.location.getY() > 1 && isRowValid(2) && !isNextLeftCellEmpty() && isNextNextLeftCellEmpty() &&
+                this.color != this.board.getPiece(new Location(this.location.getX() + colorMod, this.location.getY() - 1)).getColor()){
+            moves.add(new Move(this.location, new Location(this.location.getX() + colorMod * 2, this.location.getY() + 2),
+                      new Location(this.location.getX() + colorMod, this.location.getY() - 1)));
+        }
+
+        return moves;
```
Comment (ilyash-b) [inline src/main/java/CheckersPiece.java:44]: Hard to follow.
Maybe introduce `Direction` and/or other abstractions?
`Location` with `Direction` could give new `Location`, etc.

**Review by ilyash-b — CHANGES_REQUESTED** [review]
Finished preliminary round. Main issue: missing handling of multi-step moves.

File: src/main/java/CheckersKingPiece.java (hunk 9 above)
Comment (meiravanidjar) [reply src/main/java/CheckersKingPiece.java:12]: correct ,I wanted to discuss with you about option to have only single capture ,to eliminate the recursion here ,and this is why also I replaced intermidiates from list to single location .what do you think ?is it critical to have multi-capture moves ?

Comment (ilyash-b) [reply src/main/java/CheckersKingPiece.java:12]: > is it critical to have multi-capture moves ?

Not a pro gamer but whenever I played, multi-capture moves were always there.

**Comment by ilyash-b** [comment]
@meiravanidjar , let me know if anything is unclear, we can schedule a meeting

File: src/main/java/CheckersPiece.java (hunk 11 above)
Comment (meiravanidjar) [reply src/main/java/CheckersPiece.java]: @ilyash-b lets talk about this one.
can we schedule meeting for tomorrow?

Comment (ilyash-b) [reply src/main/java/CheckersPiece.java]: Sorry, had crazy day. Let's be in touch today to schedule (using private channels)
